services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/SdJwtSigningService.java

```diff
@@ -31,12 +48,99 @@
  */
 public class SdJwtSigningService extends SigningService<String> {
 
-    public SdJwtSigningService(KeycloakSession keycloakSession, String keyId, String algorithmType) {
+    private static final Logger LOGGER = Logger.getLogger(SdJwtSigningService.class);
+
+    private final ObjectMapper objectMapper;
+    private final SignatureSignerContext signatureSignerContext;
+    private final TimeProvider timeProvider;
+    private final String tokenType;
+    private final String hashAlgorithm;
+    private final int decoys;
+    private final List<String> visibleClaims;
+    protected final String issuerDid;
+
+    public SdJwtSigningService(KeycloakSession keycloakSession, ObjectMapper objectMapper, String keyId, String algorithmType, String tokenType, String hashAlgorithm, String issuerDid, int decoys, List<String> visibleClaims, TimeProvider timeProvider, Optional<String> kid) {
         super(keycloakSession, keyId, algorithmType);
+        this.objectMapper = objectMapper;
+        this.issuerDid = issuerDid;
+        this.timeProvider = timeProvider;
+        this.tokenType = tokenType;
+        this.hashAlgorithm = hashAlgorithm;
+        this.decoys = decoys;
+        this.visibleClaims = visibleClaims;
+        KeyWrapper signingKey = getKey(keyId, algorithmType);
+        if (signingKey == null) {
+            throw new SigningServiceException(String.format("No key for id %s and algorithm %s available.", keyId, algorithmType));
+        }
+        // set the configured kid if present.
+        if (kid.isPresent()) {
+            // we need to clone the key first, to not change the kid of the original key so that the next request still can find it.
+            signingKey = signingKey.cloneKey();
+            signingKey.setKid(keyId);
+        }
+        kid.ifPresent(signingKey::setKid);
+        SignatureProvider signatureProvider = keycloakSession.getProvider(SignatureProvider.class, algorithmType);
+        signatureSignerContext = signatureProvider.signer(signingKey);
+
+        LOGGER.debugf("Successfully initiated the SD-JWT Signing Service with algorithm %s.", algorithmType);
     }
 
     @Override
     public String signCredential(VerifiableCredential verifiableCredential) {
-        throw new UnsupportedOperationException("SD-JWT Signing is not yet supported.");
+
+        DisclosureSpec.Builder disclosureSpecBuilder = DisclosureSpec.builder();
+        CredentialSubject credentialSubject = verifiableCredential.getCredentialSubject();
+        JsonNode claimSet = objectMapper.valueToTree(credentialSubject);
```

Comment:
Question: Do we need to use `JsonNode` directly here or is it possible to use some Java class dedicated for this like we're doing in other places?

If `JsonNode` is really needed, is it at least possible to introduce constants for fields like `iss`, `nbf`, `vct`, `_sd_alg` and `jti` ? I suppose things like `iss`, `nbf`, `jti` can be probably added directly to `JsonWebToken` class as they are quite generic. For others, probably into some JWT-VC specific class?

Reply:
The JsonNode needs to be used, because the claims could be anything. However, I will introduce constants for the well-known attributes you mentioned

Reply:
@wistefan Will be good. Thanks!